jlm/llvm/frontend/LlvmInstructionConversion.cpp

```diff
@@ -886,15 +905,37 @@ convert_call_instruction(::llvm::Instruction * instruction, tacsvector_t & tacs,
     return convert_memcpy_call(i, tacs, ctx);
 
   auto ftype = i->getFunctionType();
+  auto convertedFType = ConvertFunctionType(ftype, ctx);
 
   auto arguments = create_arguments(i, tacs, ctx);
   if (ftype->isVarArg())
     arguments.push_back(create_varargs(i, tacs, ctx));
   arguments.push_back(ctx.iostate());
   arguments.push_back(ctx.memory_state());
 
-  auto fctvar = ConvertValue(i->getCalledOperand(), tacs, ctx);
-  auto call = CallOperation::create(fctvar, ConvertFunctionType(ftype, ctx), arguments);
+  const variable * callee = ConvertValueOrFunction(i->getCalledOperand(), tacs, ctx);
+  // Llvm does not distinguish between "function objects" and
+  // "pointers to functions" while we need to be precise in modelling.
+  // If the called object is a function object, then we can just
+  // feed it to the call operator directly, otherwise we have
+  // to cast it into a function object.
+  if (std::dynamic_pointer_cast<const PointerType>(callee->Type()))
+  {
+    std::unique_ptr<tac> callee_cast =
+        tac::create(PointerToFunctionOperation(convertedFType), { callee });
+    callee = callee_cast->result(0);
+    tacs.push_back(std::move(callee_cast));
+  }
+  else if (std::dynamic_pointer_cast<const FunctionType>(callee->Type()))
+  {
+    // Nothing to do.
+  }
+  else
+  {
+    throw std::runtime_error("Unexpected callee type: " + callee->Type()->debug_string());
+  }
```

Comment:
`JLM_UNREACHABLE()`, no?

Reply:
I would use "unreachable" for logic errors -- i.e. things that indicate an intrinsic problem in data structures or algorithms used, as opposed to "bad user input". But renamed for now, the distinction between these is not sharp in jlm.

Reply:
reverted this -- JLM_UNREACHABLE expects char* not std::string, so I cannot pass information about the type seen, losing information valuable for debugging. Preferring to keep std::runtime_error.